### PR TITLE
Release v3.1.2

### LINKED ISSUE #2022: Incorrect combination of implicit and explicit schemes
**Describe your setup**

preCICE Version: 3.0.0 (since #1453)

**Describe the problem**

The current method of handling compositional coupling schemes with implicit coupling schemes is the following:
1. run all explicit schemes
2. run the implicit scheme

Then rerun the implicit schemes until it achieves convergence and move to the next time-window.

The implicit scheme defines a fixed-point problem on the interface, which is accelerated using the configured acceleration schemes.
This problem changes every time window, until the implicit scheme reaches convergence.

As explicit schemes run during the first iteration along the implicit scheme, they also exchange data.
Even if the exchanged data via the explicit scheme is not used in the implicit scheme, it may change available `readData` of the solver and thus changes the fixed-point problem.
Hence, the fixed-point problem of the first iteration is different than following iterations.
This numerically doesn't make sense.

Furthermore, a serial explicit scheme where the participant with the implicit scheme runs `first` sends the data of the first iteration to the coupling partner. This data should really be the data of the converged iteration. 

**Expected behaviour**

The implicit coupling scheme should run until convergence is achieved without any interference from explicit schemes. Only when the implicit scheme converges, should the explicit schemes run.

For serial schemes, where the participant with the implicit scheme
* comes first: the data of the converged iteration should be sent to the coupling partner.
* comes second: the data of the coupling partner should be received before the implicit scheme starts iterating.

**Example**

A-B parallel-implicit; A-C serial-explicit

Top is our currently incorrect method. Bottom is the expected way of handling this.

image

**Additional context**
This means that splitting the former `CouplingScheme::advance()` into substeps is more complex as assumed.
This makes remeshing for serial coupling significantly more difficult to implement, while remeshing for parallel participants should be largely unaffected.

### PR DESCRIPTION
**Pull request by fsimonis**
## How to work with this template

* [x] assign a release manager, who takes care of the process
* [ ] assign each point below to a responsible person, before you continue. Use `@member`.

Only the release manager should update this post (even tickboxes, due to race conditions in editing). Everybody else should comment on the PR with the progress.

This release closes #2022 

## Pre-PR steps

* [x] Make sure you have the latest `main` branch locally.
* [x] Create branch `hotfix-vX.Y.Z` from `main`. If needed, `git rebase main`.
* [x] Commit fixes to the hotfix branch
* [x] Check code base w.r.t code formatting (run [`precice/tools/formatting/check-format`](https://github.com/precice/precice/blob/develop/tools/formatting/check-format)) and reformat if required (run [`precice/tools/formatting/format-all`](https://github.com/precice/precice/blob/develop/tools/formatting/format-all))
* [x] Update the list of operating systems for the package generation in `.github/workflows/release.yml`
* [x] Run `tools/releasing/bumpversion.sh MAJOR.MINOR.PATCH` to bump the version
* [x] Look over the updated `CHANGELOG.md` of the hotfix branch (all)
   * Check for merged lines
   * Add things, if necessary
   * Fix wording and tense
   * Sort the entries lexicographically
   * Extract summary
* [x] Verify the version changes in:
   * [x] `CHANGELOG`
   * [x] `CMakeLists.txt`
   * [x] `tools/releasing/packaging/debian/changelog`
* [x] Commit the version bump
* [x] Push the hotfix branch to the precice repository

## Step by step guide

* [x] Open PR from `hotfix-vX.Y.Z` to `main` (use [this template](https://github.com/precice/precice/blob/develop/tools/releasing/PULL_REQUEST_TEMPLATE/hotfix_pull_request_template.md))
* [x] Do regression tests using the hotfix branch (specific revision) _list below :arrow_down:_ (all)
* [x] Fix potential problems on the hotfix branch (all)
* [x] Reorder the commits for the version bump to be the latest. `git rebase -i main`
* [x] Approve the PR with at least two reviews (all)
* [ ] Merge PR to `main` ( use `git merge --no-ff hotfix-vX.Y.Z` )
* [ ] Tag hotfix on `main` `vX.Y.Z` and verify by running `git describe --tags`
* [ ] Merge `main` back to `develop` and verify by running `git describe --tags`
* [ ] Triple check that you haven't messed anything up. (You can always discard local changes)
* [ ] Push `main` and push the `vX.Y.Z` tag
* [ ] Push `develop`
* [ ] Wait for the release pipeline
  * [ ] [To create a new draft hotfix on GitHub](https://github.com/precice/precice/releases)
  * [ ] To automatically generate packages for the latest Debian and the two latest Ubuntu LTS versions.
* [ ] Write hotfix text
* [ ] Publish the GitHub hotfix

## Regression Tests

Use the following branches:
* precice `hotfix-v3.1.2`

Note: We run the only test affected by these changes.

| State | Success | Failure | Skipped |
| --- | --- | --- | --- |
| Write | `:o:` | `:x:` | `:fast_forward:` |
| Read | :o: | :x: | :fast_forward: |

| State | Tester | Test |
| --- | --- | --- |
| :o: | @fsimonis  | [flow-over-heated-plate-partitioned-flow](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate-partitioned-flow) fluid1-openfoam - fluid2-openfoam - solid-openfoam serial |
| :o: | @fsimonis  | [flow-over-heated-plate-partitioned-flow](https://github.com/precice/tutorials/tree/master/flow-over-heated-plate-partitioned-flow) fluid1-openfoam - fluid2-openfoam - solid-openfoam parallel |
| :o: | @uekerman | [multiple-perpendicular-flaps](https://github.com/precice/tutorials/tree/develop/multiple-perpendicular-flaps) with implicit-explicit coupling instead of multi coupling |

## Post-release

* [ ] Update version specific documentation
* [ ] Flag [Arch Linux AUR package](https://aur.archlinux.org/packages/precice) and dependants as out-of-date.
* [ ] Update Spack recipe
* [ ] Update Website:
    * [ ] Bump version in [`_config.yml`](https://github.com/precice/precice.github.io/blob/master/_config.yml)
    * [ ] Update the [XML reference](https://github.com/precice/precice.github.io/blob/master/_includes/xmlreference.md) using `binprecice md`

### Release new version for bindings (to ensure compatibility with newest preCICE version)

- [ ] [Fortran module](https://github.com/precice/fortran-module/compare/master...develop)

### Marketing

* [ ] Write on [Discourse](https://precice.discourse.group/)
* [ ] Write on [Matrix](https://matrix.to/#/#precice_lobby:gitter.im?web-instance[element.io]=app.gitter.im)

To open a new PR with this template, use this [PR template query](https://github.com/precice/precice/compare/new?template=hotfix_pull_request_template.md)

### CODE REVIEW

File: src/cplscheme/CompositionalCouplingScheme.cpp [hunk 1]
```diff
@@ -25,18 +25,61 @@ T min(T a, T b)
...
+} // namespace
+
+void CompositionalCouplingScheme::checkCompatibleTimeWindowSizes(const CouplingScheme &impl, const CouplingScheme &expl) const
+{
+  if (compatibleTimeWindowSizes(impl, expl)) {
+    return;
+  }
+  auto   local        = expl.localParticipant();
+  auto   explPartners = expl.getCouplingPartners();
+  auto   implPartners = impl.getCouplingPartners();
```
Comment (BenjaminRodenberg) [inline src/cplscheme/CompositionalCouplingScheme.cpp:53]: Why Partners and not Participants? (Not relevant for the release, but maybe a tiny refactoring for `develop`)

**Review by BenjaminRodenberg — DISMISSED** [review]
I briefly checked the compositional coupling scheme. Looks ok. Only minor remarks.

File: src/cplscheme/tests/CompositionalCouplingSchemeTest.cpp [hunk 2]
```diff
@@ -242,6 +242,7 @@ BOOST_AUTO_TEST_CASE(testDummySchemeCompositionExplicit2)
   composition.initialize(0.0, 1);
   int advances = 0;
   while (composition.isCouplingOngoing()) {
+    composition.addComputedTime(1.0);
     composition.firstSynchronization({});
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/cplscheme/tests/CompositionalCouplingSchemeTest.cpp:246]: The compositional scheme needs to know when explicit schemes have reached the end of the time window and put them on hold without transferring any data.

This needs to be handled in addComputedTime of the compositional scheme, so these tests also require calling it.

File: src/cplscheme/CompositionalCouplingScheme.cpp (hunk 1 above)
Comment (fsimonis) [reply src/cplscheme/CompositionalCouplingScheme.cpp:53]: Possibly because a coupling partner always is a participant that takes part in a scheme.

Comment (BenjaminRodenberg) [reply src/cplscheme/CompositionalCouplingScheme.cpp:53]: I'm just wondering about the wording here. Is partner and participant the same thing?

File: src/cplscheme/tests/CompositionalCouplingSchemeTest.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenjaminRodenberg) [reply src/cplscheme/tests/CompositionalCouplingSchemeTest.cpp:246]: Do the test break, if we don't call `addComputedTime(1.0)`? Did they break before the bugfix (I assume no)? I guess everything is in order the way it is, but still interesting, if the tests did not break before even though we forgot to do a call that was necessary.

[the remark this answers is not in the document]
Comment (fsimonis) [reply src/cplscheme/tests/CompositionalCouplingSchemeTest.cpp:246]: The dummy coupling scheme does not require `addComputedTime` to be called, while the compositional scheme itself now does.

So these calls weren't required before this fix.

File: src/cplscheme/CompositionalCouplingScheme.cpp (hunk 1 above)
Comment (fsimonis) [reply src/cplscheme/CompositionalCouplingScheme.cpp:53]: Not every participant in the simulation needs to be coupled to the local participant.
So, this distinction kind of makes sense. It could also just be `getCoupledParticipants()`.

**Comment by MakisH** [comment]
Just FYI, the system tests succeed: https://github.com/precice/tutorials/actions/runs/9205577625/job/25321543302

Triggered with:

```
gh workflow run run_testsuite_manual.yml -f suites=release_test -f build_args="PRECICE_REF:hotfix-v3.1.2,OPENFOAM_ADAPTER_REF:v1.3.0,PYTHON_BINDINGS_REF:v3.1.0,FENICS_ADAPTER_REF:v2.1.0,SU2_VERSION:7.5.1,SU2_ADAPTER_REF:64d4aff,TUTORIALS_REF:9d28c77" --ref=develop
```

(the `PRECICE_REF:hotfix-v3.1.2` will not catch any updates in this branch, use a specific commit, instead. With the clear issue with rebasing pointing to the same commit as HEAD.)

**Review by uekerman — DISMISSED** [review]
I mainly focused on the new integration tests. The current ones look good,  but I believe we could (easily) increase test coverage, see below.

I did some test with the [multiple perpendicular flaps](https://precice.org/tutorials-multiple-perpendicular-flaps.html) tutorial replacing the multi coupling scheme with an implicit-explicit composition. Tests produced meaningful physical and numerical results: I looked at some vtks and checked that the iteration numbers of the implicit scheme are not significantly influenced by the explicit scheme (first, second, parallel).

File: tests/serial/compositional/data/Parallel.cpp [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+#include "testing/Testing.hpp"
+
+#include <array>
+#include <precice/precice.hpp>
+
+BOOST_AUTO_TEST_SUITE(Integration)
+BOOST_AUTO_TEST_SUITE(Serial)
+BOOST_AUTO_TEST_SUITE(Compositional)
+BOOST_AUTO_TEST_SUITE(Data)
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/Parallel.cpp:11]: It's easier to run them together, prevents unnecessary prefixing of test names and allows to create a helper file for the entire suite.

File: tests/serial/compositional/data/Parallel.xml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+  <coupling-scheme:serial-explicit>
+    <participants first="IA" second="EA" />
+    <time-window-size value="1" />
+    <max-time-windows value="2" />
+    <exchange data="D-EA" mesh="M-EA" from="EA" to="IA" initialize="true" />
+    <exchange data="D-IA-EA" mesh="M-EA" from="IA" to="EA" />
+  </coupling-scheme:serial-explicit>
+
+  <coupling-scheme:serial-explicit>
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/Parallel.xml:93]: The idea is to test both versions of compositional participant goes first and goes second.

Does testing a parallel coupling scheme add any value? It's a pretty simple case.

[the remark this answers is not in the document]
Comment (uekerman) [reply tests/serial/compositional/data/Parallel.xml:93]: The tests have two explicit schemes. Currently both are serial and test the same thing. One could be serial and the other one parallel.

> Does testing a parallel coupling scheme add any value?

I would say yes. It is again a different case.

File: tests/serial/compositional/data/SerialFirst.xml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    <time-window-size value="1" />
+    <max-time-windows value="2" />
+    <exchange data="D-IA-IB" mesh="M-IB" from="IA" to="IB" />
+    <exchange data="D-IB-IA" mesh="M-IB" from="IB" to="IA" />
+    <max-iterations value="2" />
+    <min-iterations value="2" />
+  </coupling-scheme:serial-implicit>
+
+  <coupling-scheme:serial-explicit>
+    <participants first="IA" second="EA" />
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/SerialFirst.xml:86]: I try to test the same serial-explicit setup with both cases of the serial-implicit participant going first and second.

Doesn't it cover more cases to use IA first + data initialization than using IA second?

I could add an IA second version of this.

File: tests/serial/compositional/data/SerialFirst.cpp [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+#include <array>
+#include <precice/precice.hpp>
+
+BOOST_AUTO_TEST_SUITE(Integration)
+BOOST_AUTO_TEST_SUITE(Serial)
+BOOST_AUTO_TEST_SUITE(Compositional)
+BOOST_AUTO_TEST_SUITE(Data)
+BOOST_AUTO_TEST_CASE(SerialFirst)
+{
+  PRECICE_TEST("EA"_on(1_rank), "IA"_on(1_rank), "IB"_on(1_rank), "EB"_on(1_rank));
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/SerialFirst.cpp:14]: I was thinking about this, but defining these read and write lambdas makes this setup pretty minimal. It's really only putting data in and testing what's coming out.

This explicit style is pretty simple to debug.

[the remark this answers is not in the document]
Comment (uekerman) [reply tests/serial/compositional/data/SerialFirst.cpp:14]: I don't disagree. When reviewing it, my main concern was that I wanted to quickly see the differences between the tests. One could also handle this by documentation though.

File: tests/serial/compositional/data/SerialFirst.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/SerialFirst.xml:86]: > Do we expect difficulties with compositional coupling?

It's a different code path, which is good to have a test for.

File: tests/serial/compositional/data/Parallel.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/Parallel.cpp]: The tests encode the time window and iteration into the data and check the received values against expected ones.
So, Data sounds pretty fitting.

If you want to include timing, we could use `Timing` or `DataTiming`

File: tests/serial/compositional/data/Parallel.xml (hunk 4 above)
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/Parallel.xml]: Added an all parallel version.

File: tests/serial/compositional/data/SerialFirst.xml (hunk 5 above)
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/SerialFirst.xml]: Done

File: tests/serial/compositional/data/Parallel.cpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  };
+  auto read = [&](std::string dname) -> double {
+    double data;
+    p.readData(mesh, dname, {&vid, 1}, p.getMaxTimeStepSize(), {&data, 1});
+    return data;
+  };
+
+  if (context.isNamed("EA")) {
+    BOOST_REQUIRE(p.requiresInitialData());
+    write("D-EA", 0.1);
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply tests/serial/compositional/data/Parallel.cpp]: Done

File: tests/serial/compositional/data/Parallel.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (uekerman) [reply tests/serial/compositional/data/Parallel.cpp]: OK, now I understand how you meant "data". In the other tests, we don't test that correct data is communicated? If yes, keep it. We can still rename when adding more tests in the future.

**Review by uekerman — APPROVED** [review]
Test coverage looks good now. I did not check the actual results of the tests in detail though.

**Comment by fsimonis** [comment]
@uekerman Thanks

I'll rebase, request approvals again and then release.

File: src/cplscheme/BaseCouplingScheme.hpp [hunk 8]
```diff
@@ -403,6 +403,9 @@ class BaseCouplingScheme : public CouplingScheme {
   /// @copydoc cplscheme::CouplingScheme::implicitDataToReceive()
   ImplicitData implicitDataToReceive() const override;
 
+  /// @copydoc cplscheme::CouplingScheme::localParticipnat()
```
Comment (davidscn) [inline src/cplscheme/BaseCouplingScheme.hpp:406]:
```suggestion
  /// @copydoc cplscheme::CouplingScheme::localParticipant()
```

File: src/cplscheme/CompositionalCouplingScheme.cpp [hunk 9]
```diff
@@ -25,18 +25,61 @@ T min(T a, T b)
...
+  }
+  double idt = impl.getTimeWindowSize();
+  double edt = expl.getTimeWindowSize();
+  if (math::equals(edt, idt)) {
+    return true; // equal is compatible
+  }
+  if (math::smaller(edt, idt)) {
+    return false; // smaller edt is incompatible
+  }
+  return math::equals(std::remainder(edt, idt), 0.0);
```
Comment (davidscn) [inline src/cplscheme/CompositionalCouplingScheme.cpp:42]: Just as a side comment: the many returns and nested if make the code rather hard to follow. The code here could, for example, be simplified as

```suggestion
return math::equals(explTime, implTime) || math::equals(std::remainder(explTime, implTime), 0.0);
```

Not sure which of these two versions is easier to read.

File: src/cplscheme/BaseCouplingScheme.hpp (hunk 8 above)
Comment (fsimonis) [reply src/cplscheme/BaseCouplingScheme.hpp:406]: Thanks! Also fixed another one.

**Review by uekerman — APPROVED** [review]

File: src/cplscheme/CompositionalCouplingScheme.cpp (hunk 9 above)
Comment (fsimonis) [reply src/cplscheme/CompositionalCouplingScheme.cpp]: This can definitely be shortened. The verbose way is easier to document while developing.

**Review by davidscn — APPROVED** [review]

**Review by IshaanDesai — APPROVED** [review]
